runtime/src/accounts_db.rs

```diff
@@ -3691,50 +3724,52 @@ impl AccountsDb {
             .enumerate()
             .map(|(i, (meta, account))| {
                 let hash = hashes.map(|hashes| hashes[i]);
-                let cached_account =
-                    self.accounts_cache
-                        .store(slot, &meta.pubkey, (*account).clone(), hash);
+
+                let account = account
+                    .map(|account| account.to_account_shared_data())
+                    .unwrap_or_default();
+                let account_info = AccountInfo {
+                    store_id: CACHE_VIRTUAL_STORAGE_ID,
+                    offset: CACHE_VIRTUAL_OFFSET,
+                    stored_size: CACHE_VIRTUAL_STORED_SIZE,
+                    lamports: account.lamports(),
+                };
+
+                let cached_account = self.accounts_cache.store(slot, &meta.pubkey, account, hash);
                 // hash this account in the bg
                 match &self.sender_bg_hasher {
                     Some(ref sender) => {
                         let _ = sender.send(cached_account);
                     }
                     None => (),
                 };
-
-                AccountInfo {
-                    store_id: CACHE_VIRTUAL_STORAGE_ID,
-                    offset: CACHE_VIRTUAL_OFFSET,
-                    stored_size: CACHE_VIRTUAL_STORED_SIZE,
-                    lamports: account.lamports(),
-                }
+                account_info
             })
             .collect()
     }
 
     fn store_accounts_to<
         F: FnMut(Slot, usize) -> Arc<AccountStorageEntry>,
         P: Iterator<Item = u64>,
+        T: ReadableAccount,
```

Comment:
Will `T` here always be an `AccountSharedData`, or is there something else we have in mind?

Reply:
comment right above here is an example that isn't AccountSharedData.
https://github.com/solana-labs/solana/pull/16690#discussion_r621561972
Also, serialized accounts from program execution are possible. Or, something that holds Arc<AccountSharedData> internally or holds the contents of AccountSharedData in a single vector allocation from load time (more like a store or serialized for execution).

My plan was to make the T: ReadableAccount or impl ReadableAccount change after the change for Option<&AccountSharedData> I've already changed other code paths to ReadableAccount. That was the plan all along. AccountSharedData is slowly but surely having no more public members and all access is through ReadableAccount.

Reply:
sorry - this pr wasn't meant for review yet. it is a test bed for several ideas.

Reply:
steve had just asked where I was going and some 'why'. Feel free to discuss or ask other questions.